File: nnerik/day04.py
# Read input
input = open(0).read().strip().split("\n")


# Part 1
total = 0
for y in range(len(input)):
    for x in range(len(input[y])):
        for dx, dy in [(1, 0), (0, 1), (1, 1), (1, -1)]:
            if not x + 3 * dx < len(input[y]) or not 0 <= y + 3 * dy < len(input):
                continue
            word = "".join(input[y + n * dy][x + n * dx] for n in range(4))
            if word in ("XMAS", "SAMX"):
                total += 1

print("Part 1:", total)


# Part 2
total = 0
for y in range(len(input) - 2):
    for x in range(len(input[y]) - 2):
        if (
            input[y + 1][x + 1] == "A"
            and (input[y][x], input[y + 2][x + 2]) in (("M", "S"), ("S", "M"))
            and (input[y + 2][x], input[y][x + 2]) in (("M", "S"), ("S", "M"))
        ):
            total += 1

print("Part 2:", total)

File: archive/2024/fredriv/day04/day04.scala
import scala.io.Source
import scala.compiletime.ops.double

@main def day04 =
    val input = Source.fromFile("input.txt").getLines.map(_.toArray).toArray

    val horizontal = input.map(_.mkString)
    val vertical = input.transpose.map(_.mkString)
    val diagonal = diagonals(input) ++ diagonals(input.map(_.reverse))
    val total = (horizontal ++ vertical ++ diagonal).map(line =>
        line.sliding(4).count(s => s == "XMAS" || s == "SAMX")
    ).sum

    println("Part 1: " + total)
    println("Part 2: " + findXMas(input))


def findXMas(a: Array[Array[Char]]): Int =
    var total = 0
    for x <- 1 until a.length - 1
        y <- 1 until a.length - 1
    do
        if a(y)(x) == 'A' &&
            Set(a(y-1)(x-1), a(y+1)(x+1)) == Set('M', 'S') &&
            Set(a(y-1)(x+1), a(y+1)(x-1)) == Set('M', 'S')
        then total += 1

    total


def diagonals(a: Array[Array[Char]]): IndexedSeq[String] =
    val h = a.length
    val w = a(0).length
    for d <- 0 until (h + w - 1) yield
        val sb = new StringBuilder
        for x <- 0 to d do
            val y = d - x
            if y < h && x < w then
                sb.append(a(y)(x))
        sb.toString
